src/realm/sync/network/default_socket.cpp

```diff
@@ -479,8 +494,11 @@ DefaultSocketProvider::DefaultSocketProvider(const std::shared_ptr<util::Logger>
 DefaultSocketProvider::~DefaultSocketProvider()
 {
     m_logger_ptr->trace("Default event loop teardown");
-    if (m_keep_running_timer)
-        m_keep_running_timer->cancel();
+    m_service.post([this](Status status) {
+        REALM_ASSERT(status.is_ok());
+        if (m_keep_running_timer)
+            m_keep_running_timer->cancel();
+    });
```

Comment:
There is a slight chance of this not being called if it is still on the pending queue when service.stop() is called.
The current service event loop run will finish processing its current set of events and will exit before processing anything on the pending queue.

Reply:
Yeah, the problem is that it's actually not safe to cancel this timer from outside the event loop. I was getting tsan/asan errors the old way. I think also given that the sole purpose of this timer is to keep the event loop running, this is maybe not the worst slight chance we could have in the codebase. I'll keep thinking about how to improve this though.

Reply:
Food for thought: If the timer hasn't been canceled, it will fail the "orphaned" assertion when the service is destroyed... (although I'm sure you've already seen this during your testing)
https://github.com/realm/realm-core/blob/master/src/realm/sync/network/network.hpp#L1923

If we could update this part of the event loop to keep running instead of returning, then we don't need the timer anymore. I do believe some of the realm-sync-tests would need to be updated since they expect the service to stop automatically once the queue is empty.
https://github.com/realm/realm-core/blob/master/src/realm/sync/network/network.cpp#L1412-L1430

Reply:
One step at a time.

Reply:
definitely...

Reply:
Should be fixed in #6265